server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java

```diff
@@ -344,4 +372,244 @@ private static ComponentTreeWsRequest toComponentTreeWsRequest(Request request)
+
+    Map<Integer, MetricDto> metricsById = Maps.uniqueIndex(metrics, MetricDto::getId);
+    MeasureTreeQuery measureQuery = MeasureTreeQuery.builder()
+            .setStrategy(MeasureTreeQuery.Strategy.valueOf(componentTreeQuery.getStrategy().name()))
+            .setNameOrKeyQuery(componentTreeQuery.getNameOrKeyQuery())
+            .setQualifiers(componentTreeQuery.getQualifiers())
+            .setPersonId(developerId)
+            .setMetricIds(new ArrayList<>(metricsById.keySet()))
+            .build();
+
+    Table<String, MetricDto, ComponentTreeData.Measure> measuresByComponentUuidAndMetric = HashBasedTable.create(components.size(), metrics.size());
+    dbClient.measureDao().selectTreeByQuery(dbSession, baseComponent, measureQuery, result -> {
+      MeasureDto measureDto = result.getResultObject();
+      measuresByComponentUuidAndMetric.put(
+              measureDto.getComponentUuid(),
+              metricsById.get(measureDto.getMetricId()),
+              ComponentTreeData.Measure.createFromMeasureDto(measureDto));
+    });
+
+    addBestValuesToMeasures(measuresByComponentUuidAndMetric, components, metrics);
+
+    return measuresByComponentUuidAndMetric;
+  }
+
+  /**
+   * Conditions for best value measure:
+   * <ul>
+   * <li>component is a production file or test file</li>
+   * <li>metric is optimized for best value</li>
+   * </ul>
+   */
+  private static void addBestValuesToMeasures(Table<String, MetricDto, ComponentTreeData.Measure> measuresByComponentUuidAndMetric, List<ComponentDto> components,
+                                              List<MetricDto> metrics) {
+    List<MetricDtoWithBestValue> metricDtosWithBestValueMeasure = metrics.stream()
+            .filter(MetricDtoFunctions.isOptimizedForBestValue())
+            .map(new MetricDtoToMetricDtoWithBestValue())
+            .collect(MoreCollectors.toList(metrics.size()));
+    if (metricDtosWithBestValueMeasure.isEmpty()) {
+      return;
+    }
+
+    Stream<ComponentDto> componentsEligibleForBestValue = components.stream().filter(IsFileComponent.INSTANCE);
+    componentsEligibleForBestValue.forEach(component -> {
+      for (MetricDtoWithBestValue metricWithBestValue : metricDtosWithBestValueMeasure) {
+        if (measuresByComponentUuidAndMetric.get(component.uuid(), metricWithBestValue.getMetric()) == null) {
+          measuresByComponentUuidAndMetric.put(component.uuid(), metricWithBestValue.getMetric(),
+                  ComponentTreeData.Measure.createFromMeasureDto(metricWithBestValue.getBestValue()));
+        }
+      }
+    });
+  }
+
+  private static List<ComponentDto> filterComponents(List<ComponentDto> components,
+                                                     Table<String, MetricDto, ComponentTreeData.Measure> measuresByComponentUuidAndMetric, List<MetricDto> metrics, ComponentTreeRequest wsRequest) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Split this 197 characters long line (which is greater than 180 authorized). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)